### DISCUSSION TITLE
Remove marker when near location

### DISCUSSION THREAD

**Discussion by moiatto** [mapsforge/vtm#1051] [Q&A] [answered]
I cannot find a right solution to this problem:

I get a route from the OSMR provider and draw it on the map with a `PathLayer` and `destination` and `waypoints` with `ItemizedLayer` (`routePointMarkers`).
Everything works perfectly.

Now, I want to remove the waypoint markers when the location is close to the geopoint of this marker.

On location change, approximately every 1sec (range 880ms - 1100ms), I check the distance between the location and the geopoint:

```
GeoPoint p = new GeoPoint(location.getLatitude(), location.getLongitude());

if (routePointMarkers.size() > 0) {
    for (int i = 0; i < routePointMarkers.size(); i++)  {
        MarkerInterface item = (MarkerInterface) routePointMarkers.getItemList().get(i);

        if ((p.sphericalDistance(item.getPoint())) < 5) // Meters
        {
            if (i == (routePointMarkers.size()-1))
                stopNavigation();
            else
                removePoint(i);
            break;
        }
    }
}
```

The problem is the location speed (vehicle speed).
For example, setting the distance to 5 meters and considering that the position update is 1 time/sec  =  5 m/s (about 18 km/h).
if the vehicle speed is over 50 km/h, the probability of removing a point is low.

what kind of solution did you find?

**Comment by devemux86** [comment]
Is this question related to VTM map library?

It seems more a case of routing or navigation related to the application implementation.

**Comment by moiatto** [comment]
I use VTM map library.

I need to remove the marker when gps location (locationlayer) is near the marker (approx radius = 5mt). 
I tried with geopoint saved in marker but with the above problem.

A solution that uses the opengl capabilities of the map?

**Comment by devemux86** [comment, marked as answer]
`ItemizedLayer` has methods to add and remove markers.

**Comment by moiatto** [comment]
Maybe I didn't explain it clearly enough.

I already add and remove items in `ItemizedLayer`.

My problem is how remove item when location is near it (distance <= 5 meters) for the problem explained above

**Reply by devemux86** [reply]
It seems you need to implement an algorithm to track the distance.

However this is related to the app implementation,
the map libraries just render maps and overlays.

**Comment by moiatto** [comment]
Understood.

A last question about issue #578

OnLocationChanged function (gps update every 1 time/sec) I have implemented the `map.animator().animateTo(1000, mapPosition)`.
After random time, 10 sec, 1 minute or 2 minute, the same defect occurs. I have tried `map.animator().cancel();` too but nothing.

Now, I noticed that with a simple tap on the map it starts working again. If no solution, how can I replicate the tap programmatically?

**Reply by devemux86** [reply]
You can move the map programmatically, see the map.viewport() methods.

**Reply by moiatto** [reply]
Do you mean this ?

mMap.viewport().syncViewport();

**Reply by devemux86** [reply]
I do not know what you want to do.

See this class to make changes to map position.

**Reply by moiatto** [reply]
understood

**Reply by moiatto** [reply]
I looked at the class (which i already use) but i didn't find anything new compared to what i already use.

OnLocationChange (Location location) ....
```
map.viewport().getMapPosition(mapPosition);

mapPosition.setPosition(locationGps.getLatitude(), locationGps.getLongitude());
mapPosition.setBearing(-locationGps.getBearing());
mapPosition.setTilt(tiltMapView2D);

// map.animator().animateTo(1000, mapPosition);     // used before but Animation stutters after a few seconds 

map.animator().animateTo(980, mapPosition);          // GPS update every 1 sec.

map.updateMap(true);                     // is useful ?
map.viewport().syncViewport();     // is useful ?
```

**Reply by devemux86** [reply]
Cannot clearly understand what you mean.

API in all map libraries just allows to animate to map positions with a selected animation time.

Smooth movement needs interpolated map positions with very close points or a [Kalman filter](https://en.wikipedia.org/wiki/Kalman_filter).
However this is custom logic that should be implemented in the applications.

**Reply by moiatto** [reply]
ok. thanks
